### ISSUE TITLE
[Bug] DeepSeek-R1 671B FP8 emits non-finite logits (device-side assert in multinomial) on Blackwell sm_100 with DP-attention disabled, after 2026-05-16 sglang/sgl-kernel update

### ISSUE THREAD

**Issue by janbernloehr** [sgl-project/sglang#27185]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

DeepSeek-R1 671B FP8 inference on Blackwell (sm_100, e.g. GB200) with the FlashInfer TRT-LLM MoE/MLA backends and **data-parallel attention disabled** produces non-finite (NaN/Inf) logits during decode. The non-finite probabilities trip a device-side assert in sampling:

```text
/pytorch/aten/src/ATen/native/cuda/TensorCompare.cu:109: ... Assertion `probability tensor contains either inf, nan or element < 0` failed.
```

The abort surfaces from `torch.multinomial` in the sampler. The Python frame visible at abort time (for example `MinNewTokensPenalizer._apply`, `prepare_for_decode`, or `process_batch_result_prefill`) is misleading: because CUDA kernels execute asynchronously, the reported frame is wherever the next device sync happens, not where the non-finite values were produced. The actual offending op is multinomial sampling over already-non-finite logits.

This began after the `lmsysorg/sglang:dev` image was updated on 2026-05-16. Pinning the container to the 2026-05-15 build makes the failure disappear; the next published build reintroduces it. Across that pass->fail boundary, the observed component-version delta was the **sglang / sgl-kernel** update (see Environment). The failure still reproduces on current builds (sgl-kernel `0.4.3`, sglang `g83bc77661`) as of 2026-06-03.

Scope note: an earlier variant of this workload with data-parallel attention **enabled** also failed during the initial 2026-05-16 window, but it **no longer reproduces** on recent builds and is mentioned here only as historical context. The configuration that still fails is data-parallel attention **disabled**.

### Additional context

- Bisected by container digest: the 2026-05-15 last-good image passes the identical workload; the 2026-05-16 first-bad image fails. Repeated across runs.
- A related GB200 variant with DP-attention enabled also failed when this regression first appeared on 2026-05-16 (first bad build: `g688b679af`), but it recovered upstream by 2026-05-23. The last failing build we observed for that variant was in the `g16b3edc84` / `g229cadec0` range, and the first passing build we verified was `g208397aff`.
- A related GB300 variant showed the same NaN/Inf probability-tensor symptom on 2026-05-16, then recovered by 2026-05-18/19. That recovery aligns with upstream PR #25488 / commit `9869ef08` (revert of `2f817187`), so GB300 appears to have been fixed earlier by that change.

### Reproduction

Configuration that reproduces:

- Model: `deepseek-ai/DeepSeek-R1`
- Precision / path: FP8 decode path with FlashInfer TRTLLM MLA + FlashInfer TRTLLM FP8 block-scale MoE
- Hardware: Blackwell `sm_100` (for example GB200), 2 nodes x 4 GPUs per node, `TP=8`
- Data-parallel attention: **disabled** (do **not** pass `--enable-dp-attention`)

Last-known-good vs first-bad multi-arch `lmsysorg/sglang:dev` container:

- LAST-GOOD (2026-05-15):
  - arm64 `sha256:89be79a78f3bb4d66086e066b736ee15ce7f3525c0bee21bde478383490a4f31`
  - amd64 `sha256:b9c412b5a0e979c7530cc3910e62c93ff494de58c7c741f6c8ef974aa7c10cc7`
- FIRST-BAD (2026-05-16):
  - arm64 `sha256:fb6ee3204ddf3a962ab713499609b0e7c6bd77bfb7d50b53f0f702557edc1b36`
  - amd64 `sha256:34c4ca338f3d5976b52bd54a9b37cfb01b654e76f52f40fd75d65ee3ce15ea9a`

Launch on each node (`NODE_RANK=0` on the first node, `1` on the second; `HEAD_NODE` is the first node's address):

```bash
python -m sglang.launch_server \
  --model-path deepseek-ai/DeepSeek-R1 \
  --served-model-name deepseek-r1-671b \
  --tp 8 \
  --nnodes 2 \
  --node-rank ${NODE_RANK} \
  --dist-init-addr ${HEAD_NODE}:62061 \
  --mem-fraction-static 0.79 \
  --context-length 2048 \
  --piecewise-cuda-graph-max-tokens 2048 \
  --disable-radix-cache \
  --weight-loader-prefetch-checkpoints \
  --enable-nccl-nvls \
  --enable-symm-mem \
  --attention-backend trtllm_mla \
  --moe-runner-backend flashinfer_trtllm \
  --enable-flashinfer-trtllm-moe \
  --numa-node 0 0 1 1 \
  --host 0.0.0.0 \
  --port 41097
```

After the server reports ready, send a completion request:

```bash
curl -s http://${HEAD_NODE}:41097/v1/completions \
  -H 'Content-Type: application/json' \
  -d '{"model":"deepseek-r1-671b","prompt":"Hello, world!","max_tokens":16,"temperature":0.7}'
```

Expected behavior:

- HTTP 200 with a generated continuation
- Finite logits throughout sampling
- No device-side assert

Actual behavior on the bad builds:

- The first decode request aborts with

```text
/pytorch/aten/src/ATen/native/cuda/TensorCompare.cu:109: ... Assertion `probability tensor contains either inf, nan or element < 0` failed.
```

- With `CUDA_LAUNCH_BLOCKING=1`, the failure lands directly on `torch.multinomial`, confirming the sampler is consuming already-non-finite logits.

Suspected commits between last-good and first-bad, offered only as bisect starting points and not confirmed root cause:

- `54221dd9` (FP4 routed MoE `symm_output` buffer reuse)
- `ee937954` (Triton fused FP8-quantize for MLA chunked-prefill K/V)

Related symptom in a different path: [#21460](https://github.com/sgl-project/sglang/issues/21460) reports the same assert class with DP-attention + EAGLE/MTP on Hopper.

### Environment

Captured from a current failing run on 2026-06-03 (sgl-kernel `0.4.3`). `python -m sglang.check_env` is run as part of every job; the values below are taken from that output for the failing DP-attention-disabled run:

```text
sglang: 0.0.0.dev1+g83bc77661
sgl_kernel: 0.4.3
flashinfer_python: 0.6.11.post1
flashinfer_cubin: 0.6.11.post1
flashinfer_jit_cache: 0.6.11.post1+cu130
torch: 2.11.0+cu130
triton: 3.6.0
transformers: 5.8.1
torchao: 0.17.0
numpy: 2.3.5
xgrammar: 0.2.1
Python: 3.12.3
CUDA: 13.0.1
NCCL: 2.28.3-1
GPU: NVIDIA Blackwell (sm_100, compute capability 10.0)
CPU: aarch64 (Neoverse-V2)
NVIDIA driver: 580.159.04
```

Key correlation: `flashinfer-python` remained at `0.6.11.post1` across the entire pass->fail boundary, from the last-good 2026-05-15 build through the current failing 2026-06-03 build. The regression therefore correlates with the `sglang` / `sgl-kernel` update, not a `flashinfer-python` version bump. This is a correlation from the version delta and the digest A/B; it does not by itself exonerate the FlashInfer kernels invoked by sglang.

**Comment by janbernloehr** [comment]
Closing — the `multinomial` device-side assert no longer reproduces.

On current nightlies, DeepSeek-R1-671B FP8 (TP=8, **DP-attention disabled**) on Blackwell (GB200 / GB300) runs cleanly again. The crash last reproduced on `0.0.0.dev1+g66989a764` (2026-06-11) and has been gone on every build from `0.0.0.dev1+ga14d1a565` (2026-06-13) onward.

The fix is #27883, which adds the default-on `SGLANG_SANITIZE_NAN_LOGITS` guard — `Sampler._preprocess_logits()` now calls `sanitize_nan_logits()` before sampling, so non-finite logits are scrubbed before they reach `torch.multinomial` (the exact "probability tensor contains inf/nan" assert this hit). Being at the sampler entry, it's path-agnostic and covers the DP-attention-disabled case here.

Note for posterity: #27883 is a defensive sampler-level mask rather than a root-cause fix — the FP8 decode path may still emit NaN/Inf on some tokens, they're just sanitized now instead of aborting. The crash is resolved; I'll reopen if it resurfaces. Thanks!

_Posted with assistance from the `opus` AI model._
